test/extensions/filters/http/wasm/wasm_filter_test.cc

```diff
@@ -255,6 +255,51 @@ TEST_P(WasmHttpFilterTest, AsyncCall) {
   }
 }
 
+TEST_P(WasmHttpFilterTest, AsyncCallAfterDestroyed) {
+  setupConfig(TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/extensions/filters/http/wasm/test_data/async_call_cpp.wasm")));
+  setupFilter();
+
+  Http::TestHeaderMapImpl request_headers{{":path", "/"}};
+  Http::MockAsyncClientRequest request(&cluster_manager_.async_client_);
+  Http::AsyncClient::Callbacks* callbacks = nullptr;
+  EXPECT_CALL(cluster_manager_, get(Eq("cluster")));
+  EXPECT_CALL(cluster_manager_, httpAsyncClientForCluster("cluster"));
+  EXPECT_CALL(cluster_manager_.async_client_, send_(_, _, _))
+      .WillOnce(
+          Invoke([&](Http::MessagePtr& message, Http::AsyncClient::Callbacks& cb,
+                     const Http::AsyncClient::RequestOptions&) -> Http::AsyncClient::Request* {
+            EXPECT_EQ((Http::TestHeaderMapImpl{{":method", "POST"},
+                                               {":path", "/"},
+                                               {":authority", "foo"},
+                                               {"content-length", "11"}}),
+                      message->headers());
+            EXPECT_EQ((Http::TestHeaderMapImpl{{"trail", "cow"}}), *message->trailers());
+            callbacks = &cb;
+            return &request;
+          }));
+
+  EXPECT_EQ(Http::FilterHeadersStatus::StopIteration,
+            filter_->decodeHeaders(request_headers, false));
+
+  EXPECT_CALL(request, cancel()).WillOnce([&]() { callbacks = nullptr; });
+
+  // Destroy the Context, Plugin and VM.
+  filter_.reset();
+  plugin_.reset();
+  wasm_.reset();
+
+  Http::MessagePtr response_message(new Http::ResponseMessageImpl(
+      Http::HeaderMapPtr{new Http::TestHeaderMapImpl{{":status", "200"}}}));
+  response_message->body().reset(new Buffer::OwnedImpl("response"));
+
+  // (Don't) Make the callback on the destroyed VM.
+  EXPECT_EQ(callbacks, nullptr);
+  if (callbacks) {
+    callbacks->onSuccess(std::move(response_message));
+  }
+}
```

Comment:
Actually, could you add tests for all the types (HTTP callouts, gRPC callouts and gRPC streams)?

Reply:
Added a gRPC one.  We currently don't have stream tests.  Need to fix that: https://github.com/envoyproxy/envoy-wasm/issues/273

That said, these tests are almost completely useless.  This issue is a threading/config interaction issue and these tests mock out so much of the system as to essentially test nothing but that the lines I added do what they say they do.

We need integration tests for all these and I would contend we should remove the mock tests since they give a false sense of security without actually testing anything that would not be better tested by the integration tests.